File: ios/DesMoinesInsider/Services/VersionCheckService.swift
import Foundation
import Observation

/// Launch-time minimum-supported-version gate (IOS-AUDIT-REL-001).
///
/// Calls the `version-check` edge function on launch with the app's
/// `CFBundleShortVersionString`. When the running binary is below the
/// server-defined floor, `forceUpgrade` flips true and the app shell shows a
/// blocking `ForceUpdateView`. This is the client half of the CLAUDE.md
/// force-update flow that lets the backend retire an old binary before shipping
/// a backward-incompatible change.
///
/// Fails open: any network/parse error leaves `forceUpgrade == false` so a
/// backend hiccup can never lock users out of a perfectly-supported build.
@MainActor
@Observable
final class VersionCheckService {
    static let shared = VersionCheckService()

    /// True when the running version is below the server's minimum-supported
    /// floor and the app must block until the user updates.
    private(set) var forceUpgrade = false

    /// App Store URL to send the user to when an upgrade is required.
    private(set) var storeURL: URL?

    /// Copy for the blocking screen, supplied by the server.
    private(set) var message = "Please update to the latest version to continue."

    /// Latest available version, when the server reports an optional update.
    private(set) var latestVersion: String?

    private let supabase = SupabaseService.shared.client

    private init() {}

    struct Response: Decodable {
        let platform: String
        let currentVersion: String
        let minSupportedVersion: String
        let latestVersion: String
        let forceUpgrade: Bool
        let updateAvailable: Bool
        let storeUrl: String
        let message: String
    }

    private struct Payload: Encodable {
        let platform: String
        let version: String
    }

    /// Performs the launch check. Safe to call unauthenticated; no-ops under UI
    /// tests so screenshots/automation never hit the gate.
    func checkOnLaunch() async {
        guard !Config.isUITesting, let client = supabase else { return }

        do {
            let response: Response = try await client.functions.invoke(
                "version-check",
                options: .init(body: Payload(platform: "ios", version: Config.appVersion))
            )
            forceUpgrade = response.forceUpgrade
            storeURL = URL(string: response.storeUrl)
            message = response.message
            latestVersion = response.updateAvailable ? response.latestVersion : nil
            #if DEBUG
            AppLogger.network.info(
                "version-check: \(Config.appVersion) min=\(response.minSupportedVersion) force=\(response.forceUpgrade)"
            )
            #endif
        } catch {
            // Fail open — never block a supported build because the check failed.
            #if DEBUG
            AppLogger.network.warning("version-check failed (failing open): \(error.localizedDescription)")
            #endif
        }
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/util/VersionCheckService.kt
package com.desmoines.aipulse.util

import com.desmoines.aipulse.BuildConfig
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.functions.functions
import io.ktor.client.statement.bodyAsText
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import javax.inject.Inject
import javax.inject.Singleton

/** Parsed `version-check` response. Additive — unknown keys ignored. */
@Serializable
data class VersionCheckResult(
    val platform: String = "android",
    val currentVersion: String = "",
    val minSupportedVersion: String = "",
    val latestVersion: String = "",
    val forceUpgrade: Boolean = false,
    val updateAvailable: Boolean = false,
    val storeUrl: String = DEFAULT_STORE_URL,
    val message: String = "",
) {
    companion object {
        const val DEFAULT_STORE_URL =
            "https://play.google.com/store/apps/details?id=com.desmoines.aipulse"
    }
}

/**
 * Launch-time minimum-version gate. Mirrors iOS VersionCheckService.swift.
 *
 * Calls the public `version-check` edge function with the running platform + version.
 * Fails open: any network/parse error leaves [state] non-blocking, so the app is never
 * stuck behind a transient backend issue. When forceUpgrade is true the UI shows a
 * blocking ForceUpdateScreen.
 */
@Singleton
class VersionCheckService @Inject constructor(
    private val client: SupabaseClient?,
    private val json: Json,
) {
    private val _state = MutableStateFlow(VersionCheckResult())
    val state: StateFlow<VersionCheckResult> = _state.asStateFlow()

    /** Query the backend once. Safe to call on launch; never throws. */
    suspend fun check() {
        val supabase = client ?: return
        try {
            val payload = buildJsonObject {
                put("platform", "android")
                put("version", BuildConfig.VERSION_NAME)
            }
            val body = supabase.functions("version-check", body = payload).bodyAsText()
            _state.value = json.decodeFromString(VersionCheckResult.serializer(), body)
        } catch (error: Throwable) {
            // Fail open — leave the default non-blocking state.
            AppLogger.network.warning("version-check failed (failing open): ${error.message}")
        }
    }
}
